build-scripts/multidirectory_test.sh

```diff
@@ -0,0 +1,229 @@
+test7() {
+    create_test_repos
+    branches build/ose cluster/dev-preview-int/install
+    expected_branches master
+    execute_test
+}
+
+# Multiple branches without jobs, multiple jobs without branches.
+test8() {
+    create_test_repos
+    branches build/ose cluster/dev-preview-int/install
+    jobs build/ose-pipeline cluster/cicd-test
+    expected_branches build/ose-pipeline cluster/cicd-test master
+    execute_test
+}
+
+# Jobs with up-to-date branches.
+test9() {
+    local shas
+    create_test_repos
+    jobs build/ose cluster/dev-preview-int/install
+    branches build/ose cluster/dev-preview-int/install
+    expected_branches build/ose cluster/dev-preview-int/install master
+    shas=$(cd setup/ \
+        && git rev-parse build/ose cluster/dev-preview-int/install)
+    execute_test
+    echo "$shas" > expected_shas.txt
+    (
+        cd origin.git
+        git rev-parse build/ose cluster/dev-preview-int/install \
+            | diff -u ../expected_shas.txt -
+    )
+}
+
+# Jobs with not-up-to-date branches.
+test10() {
+    local shas
+    create_test_repos
+    branches build/ose cluster/dev-preview-int/install
+    shas=$(cd setup/ \
+        && git rev-parse build/ose cluster/dev-preview-int/install)
+    jobs build/ose cluster/dev-preview-int/install
+    expected_branches build/ose cluster/dev-preview-int/install master
+    execute_test
+    echo "$shas" > unexpected_shas.txt
+    (
+        cd origin.git
+        git rev-parse build/ose cluster/dev-preview-int/install \
+            | { grep -xf ../unexpected_shas.txt - || exit 0 && exit 1; }
+    )
+}
+
+# Extra files in the repository.
+test11() {
+    create_test_repos
+    branches build/ose
+    jobs build/ose cluster/dev-preview-int
+    expected_branches build/ose cluster/dev-preview-int master
+    (
+        cd setup/ \
+            && mkdir build-scripts/ \
+            && > build-scripts/multidirectory.sh \
+            && git add build-scripts/ \
+            && git commit --quiet --message 'extra files'
+    )
+    execute_test
+}
+
+# Rerun.
+test12() {
+    create_test_repos
+    jobs build/ose cluster/dev-preview-int
+    expected_branches build/ose cluster/dev-preview-int master
+    execute_test
+    git ls-remote --heads "$PWD/origin.git/" > expected_output.txt
+    rm -rf setup/ work/ check/
+    git clone --quiet origin.git/ setup/
+    execute_test
+    git ls-remote --heads "$PWD/origin.git/" \
+        | diff -u expected_output.txt -
+}
+
+MULTIDIRECTORY_SH=$(realpath "$(dirname "$0")/multidirectory.sh")
+GIT_AUTHOR_NAME=test
+GIT_AUTHOR_EMAIL=test@example.com
+GIT_COMMITTER_NAME=$GIT_AUTHOR_NAME
+GIT_COMMITTER_EMAIL=$GIT_AUTHOR_EMAIL
+export GIT_AUTHOR_NAME GIT_AUTHOR_EMAIL GIT_COMMITTER_NAME GIT_COMMITTER_EMAIL
+test_dir=$PWD/tests
+rm -rf "$test_dir"
+[ "$#" -gt 0 ] || set -- $(seq -f 'test%.0f' 0 12)
+for x; do
+    echo "$x"
+    mkdir -p "$test_dir/$x"
+    set +e
+    (set -e; cd "$test_dir/$x"; "$x")
+    [ "$?" -eq 0 ] || failed=1
```

Comment:
These lines are tricky.  The original was:

```
(cd "$test_dir/" && "$x") || failed=1
```

but that taught me that when `set -e` is in effect, not only an error is not
generated if any command but the last in a chain of `&&` and/or `||` fails
(which I already knew) but it also cancels the effect of `set -e` for all those
commands.

So here, because `"$x"` (the test we are running) is on the left side of `||`,
`set -e` is disable when executing it.  That led to this more complicated code,
where we need to `set +e` on the parent shell, start a subshell, `set -e`,
execute the test, test the result on the parent, and then `set -e` again.

Not the prettiest thing in the world (as any test suite written in shell is
probably going to be), but not *that* bad.